library/java/org/chromium/net/impl/CancelProofEnvoyStream.java

```diff
@@ -0,0 +1,114 @@
+package org.chromium.net.impl;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import io.envoyproxy.envoymobile.engine.EnvoyHTTPStream;
+
+/**
+ * Consistency layer above the {@link EnvoyHTTPStream} preventing unwarranted Stream operations
+ * after a "cancel" operation. There are no "synchronized" - this is Compare And Swap based logic.
+ *
+ * <p>This contraption ensures that once a "cancel" operation is invoked, there will be no further
+ * operations allowed with the EnvoyHTTPStream - subsequent operations will be ignored silently.
+ * However, in the event that that one or more EnvoyHTTPStream operations are currently being
+ * executed, the "cancel" operation gets postponed: the last concurrent operation will invoke
+ * "cancel" at the end.
+ *
+ * <p>Instance of this class start with a state of "Busy Starting". This ensure that if a cancel
+ * is invoked while the stream is being created, that cancel will be executed only once the stream
+ * is completely initialized. Doing otherwise leads to unpredictable outcomes.
+ */
+final class CancelProofEnvoyStream {
+
+  private static final int CANCEL_BIT = 0x8000;
+  private final AtomicInteger mState = new AtomicInteger(1); // Busy starting.
+  private volatile EnvoyHTTPStream mStream;                  // Cancel can come from any Thread.
+
+  /** Sets the stream. Can only be invoked once. */
+  void setStream(EnvoyHTTPStream stream) {
+    assert mStream == null;
+    mStream = stream;
+    if (cancelNeedsToBeInvoked()) {
+      mStream.cancel(); // Cancel was called meanwhile, so now this is honored.
```

Comment:
Changed to "awaitingCancel".

Reply:
Thanks for the rename! They're super verbose but very clear.

Though this comment was requesting "previously" instead of "meanwhile" in the comment.